doc/architecture/decisions/0027-recovery-rollback-strategy.md

```diff
@@ -0,0 +1,97 @@
+# Recovering from application releases
+
+27. Recovery Rollback Strategy
+
+[Next >>](9999-end.md)
+
+Date: 2023-06-16
+
+## Status
+Accepted
+
+## Context
+
+<table>
+  <tr>
+   <td>Description
+   </td>
+   <td>This ADR covers our approach for dealing with a deployment that need to be rolled-back due to application errors.
+   </td>
+  </tr>
+  <tr>
+   <td>Collaborators
+   </td>
+   <td><em>Michael Willis, Oskar Williams</em>
+   </td>
+  </tr>
+  <tr>
+   <td>Date
+   </td>
+   <td><em>16/6/23</em>
+   </td>
+  </tr>
+  <tr>
+   <td>Status
+   </td>
+   <td><em>Accepted</em>
+   </td>
+  </tr>
+</table>
+
+
+# Context
+
+When deploying a new version of software, it's important to minimise the risk of downtime, errors, and user dissatisfaction. But sometimes things go wrong, and it's necessary to have a plan to revert to a previous state.
+
+Running a live service requires regular patching, bug fixes and feature improvements.  Ultimately this means that sometimes a problem with the running service can occur as a result of a new release.  Mitigation of these issues can be achieved with good test coverage and end to end smoke testing - but there is no substitute for a live environment highlighting something new or misconfigured.
+
+This document looks at the approach to perform this action.  Two common strategies to do this are deployment **rollback** and **blue/green** deployment.
+
+
+### Blue/green deployment
+Blue/green deployment is a more advanced way to switch between versions. It means having two identical environments, one with the current version (blue) and one with the new version (green). You can test the green environment before making it live, and then switch the traffic from blue to green with a load balancer or a DNS change.
+
+#### Advantages
+Zero-downtime deployment and easy rollback
+
+#### Disadvantages
+Requires more resources, coordination, and synchronisation.
+
+### Deployment rollback
+Deployment rollback is the simplest way to undo a deployment. It means restoring the previous version of your software from a backup or a source control repository.
+
+#### Advantages
+Fast and easy to execute.
+
+#### Disadvantages
+Can cause data loss, inconsistency, or corruption if the new version has made any changes to the database or other external resources.
+
+
+# Decision
+Although larger and more complex systems may benefit from blue/green deployment due to its reduced risk of errors and data issues, more frequent and urgent deployments favour deployment rollback, as it is quicker and simpler to execute. The simplicity of rollback was chosen as the approach.
```

Comment:
It's worth noting that we _do_ have blue/green deployment, with automated smoke tests. We could easily configure this for something other than all at once deployment: https://github.com/alphagov/gdx-data-share-poc/blob/main/terraform/modules/data_share_service/codedeploy.tf#L10.

We could also configure the blue instances to stick around for faster rollback: https://github.com/alphagov/gdx-data-share-poc/blob/main/terraform/modules/data_share_service/codedeploy.tf#L23

I think the approach here is correct for once the blue/green phase is over, or for db migration issues